lambda/shared/src/main/scala/feral/lambda/tracing/TracedLambda.scala

```diff
@@ -0,0 +1,90 @@
+/*
+ * Copyright 2021 Typelevel
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package feral.lambda.tracing
+
+import cats._
+import cats.data._
+import cats.effect.std.Random
+import cats.effect.{Trace => _, _}
+import natchez._
+import feral.lambda._
+import natchez.noop.NoopSpan
+import natchez.xray.{XRay, XRayEnvironment}
+
+abstract class TracedLambda[F[_]: MonadCancelThrow, G[_], Event] {
+  def extractKernel(event: Event, context: Context[F]): F[Kernel]
```

Comment:
I wonder if we can make this a typeclass in terms of `Event`. The http one can have something special that extracts the kernel from the headers, and the rest can use a low-priority fallback.

Reply:
I added an explicit `extractKernel: Kleisli[F, (Event, Context[F]), Kernel]` parameter instead of making it a typeclass, because the instances will vary depending on what backend is being used.

Reply:
Hmm, yeah I was afraid that it might depend on the backend, too bad. 👍

Reply:
So I looked into this again and I'm not sure. I think we should toss all the headers from the event into the `Kernel` and leave it to the entrypoint to sort it out. There's nothing backend-specific about that :)